src/main/java/org/sonarlint/intellij/core/ServerIssueUpdater.java

```diff
@@ -79,19 +83,37 @@ public void fetchAndMatchServerIssues(Collection<VirtualFile> virtualFiles) {
     SonarQubeServer server = projectBindingManager.getSonarQubeServer();
     ConnectedSonarLintEngine engine = projectBindingManager.getConnectedEngine();
     String moduleKey = projectSettings.getProjectKey();
+    List<Future<Void>> updateTasks = new LinkedList<>();
 
+    // submit tasks
     for (VirtualFile virtualFile : virtualFiles) {
       String relativePath = SonarLintUtils.getRelativePath(myProject, virtualFile);
-      fetchAndMatchServerIssues(virtualFile, server, engine, moduleKey, relativePath);
+      updateTasks.add(fetchAndMatchServerIssues(virtualFile, server, engine, moduleKey, relativePath));
+    }
+
+    if (modal) {
+      waitForTasks(updateTasks);
+    }
+  }
+
+  private void waitForTasks(List<Future<Void>> updateTasks) {
+    for(Future<Void> f : updateTasks) {
+      try {
+        f.get(5, TimeUnit.SECONDS);
+      } catch (Exception ex) {
+        LOGGER.warn("ServerIssueUpdater task expired", ex);
+      }
```

Comment:
I'm wondering if this could be a good use case for [`ExecutorService.invokeAll`](http://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html#invokeAll-java.util.Collection-long-java.util.concurrent.TimeUnit-). Because it seems a bit unfair that some of the futures get more time to complete than others. (For example in case of 3 futures, the first one has 5 seconds to complete, but the last one has effectively 15.)